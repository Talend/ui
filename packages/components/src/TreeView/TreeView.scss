// TreeView styles

.tc-treeview {
	.tc-treeview-header {
		display: flex;
		justify-content: space-between;
		align-items: center;

		color: $brand-primary;
		line-height: 6rem;
		font-weight: bold;
		text-transform: uppercase;

		padding-left: $padding-large;
		border-bottom: 1px rgba(0, 0, 0, 0.12) solid;

		:global(.btn-link .tc-svg-icon) {
			stroke: $brand-primary;
		}
	}

	.tc-treeview-list {
		padding: $padding-small 0;
		overflow: auto;
		list-style-type: none;
	}
}
