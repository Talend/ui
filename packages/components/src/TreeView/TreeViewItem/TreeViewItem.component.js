import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

import TooltipTrigger from '../../TooltipTrigger';
import { Action } from '../../Actions';
import Icon from '../../Icon';
import Badge from '../../Badge';

import css from './TreeViewItem.scss';

const BASE_PADDING = 30;
const CARET_WIDTH = 12;
const PADDING = 20;

/**
 * return the default open or closed folder icon if non is specified on item
 * or if it is specified return the specified icon with `-closed` append if not opened
 * @param {String} iconName - icon name
 * @param {Boolean} isOpened - state of the item
 * @return {String}
 */
export function getItemIcon(iconName = 'talend-folder', isOpened) {
	return isOpened ? iconName : `${iconName}-closed`;
}

/**
 * Internal: you should not use it
 * Instantiate an icon based on the icon config
 *
 * @param icon The icon name of the Icon props
 * @param isOpened if the treeview is opened
 */
function TreeViewIcon({ icon, isOpened }) {
	if (typeof icon === 'object') {
		return icon.tooltipLabel ? (
			<TooltipTrigger label={icon.tooltipLabel} tooltipPlacement={icon.tooltipPlacement || 'top'}>
				<span>
					<Icon name={icon.name} className={classNames(css['tc-treeview-img'], icon.className)} />
				</span>
			</TooltipTrigger>
		) : (
			<Icon {...icon} className={classNames(css['tc-treeview-img'], icon.className)} />
		);
	}

	return (
		<Icon name={getItemIcon(icon, isOpened)} className={classNames(css['tc-treeview-folder'])} />
	);
}
TreeViewIcon.propTypes = {
	icon: PropTypes.oneOfType([PropTypes.string, PropTypes.shape(Icon.propTypes)]),
	isOpened: PropTypes.bool,
};

/**
 * Internal: you should not use it
 * Single item of TreeView component
 *
 * @param id, for qa purposes
 * @param item required, item to display
 * 		  item.actions optional, array with actions' to be displayed meta-info
 * @param onSelect required, callback function to trigger once item was clicked
 * @param onClick required, callback function to trigger once item was clicked
 *
 * @returns XML, jsx to display
 */

class TreeViewItem extends React.Component {
	static propTypes = {
		id: PropTypes.string.isRequired,
		index: PropTypes.number.isRequired,
		item: PropTypes.shape({
			id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
			name: PropTypes.string.isRequired,
			isOpened: PropTypes.bool,
			children: PropTypes.arrayOf(PropTypes.object),
			icon: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
			actions: PropTypes.arrayOf(
				PropTypes.shape({
					action: PropTypes.func,
					label: PropTypes.string,
					icon: PropTypes.string,
				}),
			),
			counter: PropTypes.number,
			showCounter: PropTypes.bool,
		}).isRequired,
		siblings: PropTypes.array,
		level: PropTypes.number.isRequired,
		onKeyDown: PropTypes.func.isRequired,
		onToggle: PropTypes.func.isRequired,
		onSelect: PropTypes.func.isRequired,
		selectedId: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
	};

	static defaultProps = {
		level: 1,
	};

	constructor(props) {
		super(props);
		this.renderIconAction = this.renderIconAction.bind(this);
		this.renderTreeViewChildren = this.renderTreeViewChildren.bind(this);
		this.onMouseLeave = this.onMouseLeave.bind(this);
		this.onMouseEnter = this.onMouseEnter.bind(this);
		this.state = {
			hovered: false,
		};
	}

	onMouseEnter() {
		this.setState({
			hovered: true,
		});
	}

	onMouseLeave() {
		this.setState({
			hovered: false,
		});
	}

	getTabIndex() {
		let shouldBeFocusable;
		if (this.props.selectedId === undefined) {
			shouldBeFocusable = this.props.level === 1 && this.props.index === 1;
		} else {
			shouldBeFocusable = this.isSelected();
		}

		return shouldBeFocusable ? 0 : -1;
	}

	hasChildren() {
		return this.props.item.children && this.props.item.children.length;
	}

	isSelected() {
		const { item, selectedId } = this.props;
		if (selectedId === undefined) {
			return false;
		}
		if (Array.isArray(selectedId)) {
			return selectedId.includes(item.id);
		}
		return item.id === selectedId;
	}

	isOpened() {
		const { children = [], isOpened = false } = this.props.item;
		if (!children.length) {
			return undefined;
		}
		return isOpened;
	}

	renderTreeViewChildren() {
		if (!this.isOpened()) {
			return null;
		}

		const { children } = this.props.item;
		return (
			<ul key="children" role="group" className={css['tc-treeview-ul']}>
				{children.map((child, i) => (
					<TreeViewItem
						id={this.props.id && `${this.props.id}-${i}`}
						item={child}
						siblings={children}
						onKeyDown={this.props.onKeyDown}
						onSelect={this.props.onSelect}
						onToggle={this.props.onToggle}
						key={i}
						index={i + 1}
						selectedId={this.props.selectedId}
						level={this.props.level + 1}
					/>
				))}
			</ul>
		);
	}

	renderIconAction({ action, id, ...actionProps }) {
		let safeId = id;
		if (!id && this.props.id) {
			safeId = `${this.props.id}-${actionProps.icon}`;
		}
		return (
			<Action
				{...actionProps}
				onClick={event => {
					event.stopPropagation();
					action(this.props.item);
				}}
				tooltipPlacement="right"
				hideLabel
				key={actionProps.label}
				id={safeId}
				link
			/>
		);
	}

	render() {
		const { id, index, item, level, onKeyDown, onSelect, onToggle, siblings } = this.props;
		const {
			isOpened = false,
			hidden,
			name,
			children = [],
			showCounter,
			actions,
			icon,
			counter = children.length,
		} = item;
		const paddingLeft = `${(level - 1) * (PADDING + CARET_WIDTH) + BASE_PADDING}px`;
		const showOpenedFolder = !!(children.length && (isOpened || this.state.hovered));

		return (
			<li // eslint-disable-line jsx-a11y/no-static-element-interactions
				id={id}
				role="treeitem"
				tabIndex={this.getTabIndex()}
				aria-expanded={this.isOpened()}
				aria-level={level}
				aria-posinset={index}
				aria-setsize={siblings.length}
				aria-selected={this.isSelected()}
				className={classNames('tc-treeview-item-li', css['tc-treeview-li'])}
				onClick={e => {
					e.stopPropagation();
					return onSelect(e, item);
				}}
				onKeyDown={e =>
					onKeyDown(e, this.containerRef, {
						...item,
						hasChildren: children.length,
						isOpened,
						siblings,
					})
				}
				data-hidden={hidden}
				onMouseEnter={this.onMouseEnter}
				onMouseLeave={this.onMouseLeave}
				ref={ref => {
					this.containerRef = ref;
				}}
			>
				<div
					className={classNames('tc-treeview-item', css['tc-treeview-item'])}
					style={{ paddingLeft }}
				>
					{children.length ? (
						<Action
							key="toggle"
							className={css['tc-treeview-toggle']}
							icon="talend-caret-down"
							iconTransform={isOpened ? undefined : 'rotate-270'}
							id={id && `${id}-toggle`}
							onClick={e => {
								e.stopPropagation();
								return onToggle(e, item);
							}}
							label=""
							aria-hidden
							tabIndex="-1"
							link
						/>
					) : null}
					<TreeViewIcon key="icon" icon={icon} isOpened={showOpenedFolder} />
					<span
						key="label"
						className={classNames('tc-treeview-item-name', css['tc-treeview-item-name'])}
					>
						{name}
					</span>
					<div key="actions" className={css['tc-treeview-item-ctrl']}>
						{showCounter && <Badge key="badge" label={counter.toString()} />}
						{actions && actions.map(this.renderIconAction)}
					</div>
				</div>
				{this.renderTreeViewChildren()}
			</li>
		);
	}
}

export default TreeViewItem;
