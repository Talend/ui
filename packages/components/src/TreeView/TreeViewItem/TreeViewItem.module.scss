// TreeView styles
$item-height: 4rem;

.tc-treeview-li {
	display: flex;
	align-items: center;
	flex-wrap: wrap;

	&:focus {
		outline: none;
	}

	:global(.btn-link) {
		padding: 0;
	}

	.tc-treeview-item {
		position: relative;
		height: $item-height;
		cursor: pointer;
		padding-right: $padding-large;

		display: flex;
		align-items: center;
		flex-grow: 1;

		transition: 200ms ease-out;

		&.disabled {
			opacity: 0.54;
			cursor: not-allowed;

			.tc-treeview-item-name {
				color: $black;
			}

			svg {
				fill: rgba($black, 0.25);
			}
		}

		.tc-treeview-toggle {
			position: absolute;
			padding: 1rem;
			display: inline-flex;
			min-height: auto;
			height: 100%;
			margin-left: -3rem;
			top: 0;

			svg {
				width: $svg-sm-size;
				fill: rgba($black, 0.6);
				margin: 0;
			}

			&:hover svg {
				fill: rgba($black, 0.7);
			}
		}

		.tc-treeview-folder {
			width: $svg-md-size;
			height: $svg-md-size;
			margin-right: $padding-large;
			flex-shrink: 0;
			flex-grow: 0;
			fill: rgba($black, 0.25);
		}

		.tc-treeview-img {
			max-height: 2rem;
			margin-right: 1rem;
			vertical-align: middle;
		}

		.tc-treeview-item-name {
			white-space: nowrap;
			overflow: hidden;
			text-overflow: ellipsis;
			max-width: 40rem;
		}
	}

	&[aria-selected='true'] > .tc-treeview-item {
		background: rgba($scooter, 0.2);
		font-weight: 600;
		color: $black;

		.tc-treeview-folder {
			fill: $brand-primary;
		}
	}

	.tc-treeview-item:hover,
	&:focus > .tc-treeview-item {
		background: $wild-sand;
		color: $scooter;

		.tc-treeview-folder {
			fill: $brand-primary;
		}

		.tc-treeview-toggle {
			svg {
				fill: $scooter;
			}
		}

		:global(.btn-link) {
			visibility: visible;
		}

		&.disabled {
			background: none;
		}
	}

	.tc-treeview-item-ctrl {
		margin-left: auto;
		display: flex;
		align-items: center;

		:global(.btn-link) {
			visibility: hidden;
			padding: 0 $padding-smaller;

			:global(.tc-svg-icon) {
				fill: rgba($black, 0.6);
				margin: 0;

				&:hover {
					fill: rgba($black, 0.7);
				}
			}
		}

		:global(.tc-badge-button) {
			height: 2rem;
			margin: 0 $padding-normal 0 0;
			padding: 0 $padding-smaller;
			border-radius: $border-radius-small;
		}
	}

	.tc-treeview-ul {
		width: 100%;
		padding: 0;
	}

	&[data-hidden='true'] {
		height: 0;
		display: none;
	}
}
