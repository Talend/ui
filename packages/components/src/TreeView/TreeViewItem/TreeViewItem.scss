// TreeView styles
$item-height: 40px;

.tc-treeview-li {
	display: flex;
	align-items: center;
	flex-wrap: wrap;

	.tc-treeview-item {
		height: $item-height;
		line-height: $item-height - 2 * $padding-small;
		cursor: pointer;
		padding: $padding-small $padding-large $padding-small 0;
		border: none;
		background: none;
		text-align: justify;
		display: flex;
		align-items: center;
		transition: 200ms ease-out;
		flex-grow: 1;

		:global(.btn-link) {
			padding: 0;
			border: none;
			display: block;

			:global(.tc-svg-icon) {
				margin-right: 0;
			}
		}

		.tc-treeview-folder {
			width: $svg-md-size;
			height: $item-height - 2 * $padding-small;
			margin-right: $padding-large;
			flex-shrink: 0;
			flex-grow: 0;
		}

		.tc-treeview-img {
			max-height: 2rem;
			margin-right: 10px;
		}

		&:hover {
			background: rgba($black, 0.05);
			color: $scooter;

			.tc-treeview-folder {
				fill: $lightning-yellow;
			}

			:global(.btn-link) {
				visibility: visible;
			}

			.tc-treeview-folder {
				&-opened {
					display: block;
				}
				&-closed {
					display: none;
				}
			}
		}

		.tc-treeview-toggle {
			width: $svg-sm-size;
			margin: 0;
			margin-right: $padding-large;
			position: relative;
			fill: rgba($black, 0.6);

			&:hover {
				fill: rgba($black, 0.7);
			}
		}

		&[data-selected='true'] {
			background: rgba($black, 0.05);
			font-weight: bold;
			color: black;

			.tc-treeview-folder {
				fill: $lightning-yellow;
			}
		}
	}

	.tc-treeview-item-ctrl {
		margin-left: auto;
		display: flex;
		align-items: center;

		:global(.btn-link) {
			margin-left: auto;
			visibility: hidden;
			padding: 0 $padding-smaller;
			line-height: unset;

			:global(.tc-svg-icon) {
				width: $svg-md-size;
				height: $svg-md-size;
				fill: rgba($black, 0.6);
				margin: 0;

				&:hover {
					fill: rgba($black, 0.7);
				}
			}
		}

		:global(.tc-badge-button) {
			height: $item-height - 2 * $padding-small;
			margin: 0 $padding-normal 0 0;
			border: none;
			padding: 0 $padding-smaller;
			border-radius: $border-radius-small;
		}
	}

	.tc-treeview-ul {
		width: 100%;
		padding: 0;
	}

	&[data-hidden='true'] {
		height: 0;
		display: none;
	}

	:global(.tc-svg-icon) {
		fill: rgba($black, 0.25);
	}
}
