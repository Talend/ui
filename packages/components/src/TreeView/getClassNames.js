import classNames from 'classnames';

const getClassName = theme => className => classNames(className, theme[className]);
export default getClassName;
