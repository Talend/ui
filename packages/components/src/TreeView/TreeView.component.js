import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import { Action } from '../';
import TreeViewItem from './TreeViewItem/';

import theme from './TreeView.scss';
import withTreeGesture from '../Tree/withTreeGesture';

/**
 * A view component to display any tree structure, like folders or categories.
 *
 * @param id, for qa purposes
 * @param structure optional, tree structure to display, see example below
 * @param headerText optional, specifies text in component's header
 * @param addAction optional, defines if 'add' button is displayed and
 *        specifies button click event callback
 * @param addActionLabel optional, specifies tooltip label for 'add' button
 * @param itemSelectCallback required, tree item click event callback function
 * @param itemToggleCallback required, tree item expand/collapse event callback function
 *
 * const defaultProps = {
 * 	structure: [{
 * 		name: 'grandpa',
 * 		children: [
 * 			{
 * 				name: 'mami',
 * 				toggled: true,
 * 				children: [
 * 					{ name: 'me', selected: true },
 * 					{ name: 'bro' },
 * 				],
 * 			},
 * 			{ name: 'aunt', toggled: false, children: [{ name: 'cousin' }] },
 * 		],
 * 		toggled: true
 * 	}],
 * 	headerText: 'some elements',
 * 	addAction: () => null,
 * 	addActionLabel: 'add element',
 * 	itemSelectCallback: () => null,
 * 	itemToggleCallback: () => null,
 * }
 *
 * <TreeView {...defaultProps} />
 *
 */
function TreeView(props) {
	const {
		id,
		headerText,
		structure,
		addAction,
		addActionLabel,
		onKeyDown,
		onSelect,
		onToggle,
		noHeader,
		className,
		selectedId,
		style,
	} = props;
	const titleId = id && `${id}-title`;
	return (
		<div className={classNames('tc-treeview', theme['tc-treeview'], className)} style={style}>
			<header className={classNames(theme['tc-treeview-header'], { 'sr-only': noHeader })}>
				<span id={titleId}>{headerText}</span>
				{addAction && (
					<Action
						label={addActionLabel}
						icon="talend-plus"
						onClick={addAction}
						tooltipPlacement="right"
						hideLabel
						link
						id={id && `${id}-add`}
						key={addActionLabel}
					/>
				)}
			</header>
			<ul className={theme['tc-treeview-list']} role="tree" aria-labelledby={titleId}>
				{structure.map((item, i) => (
					<TreeViewItem
						id={id && `${id}-${i}`}
						item={item}
						siblings={structure}
						onKeyDown={onKeyDown}
						onSelect={onSelect}
						onToggle={onToggle}
						key={i}
						index={i + 1}
						selectedId={selectedId}
						level={1}
					/>
				))}
			</ul>
		</div>
	);
}

TreeView.displayName = 'TreeView';

TreeView.defaultProps = {
	id: 'tc-treeview',
	addActionLabel: 'Add folder',
	headerText: 'Folders',
};

if (process.env.NODE_ENV !== 'production') {
	TreeView.propTypes = {
		id: PropTypes.string.isRequired,
		headerText: PropTypes.string,
		structure: PropTypes.arrayOf(TreeViewItem.propTypes.item),
		addAction: PropTypes.func,
		addActionLabel: PropTypes.string,
		onKeyDown: PropTypes.func.isRequired,
		onToggle: PropTypes.func.isRequired,
		onSelect: PropTypes.func.isRequired,
		noHeader: PropTypes.bool,
		className: PropTypes.string,
		selectedId: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array]),
		style: PropTypes.object,
	};
}

export default withTreeGesture(TreeView);
