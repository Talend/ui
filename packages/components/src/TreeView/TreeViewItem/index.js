import TreeViewItem from './TreeViewItem.component';

export default TreeViewItem;
