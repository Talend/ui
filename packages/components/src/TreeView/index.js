import TreeView from './TreeView.component';

export default TreeView;
