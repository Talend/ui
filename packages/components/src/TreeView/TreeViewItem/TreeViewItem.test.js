import React from 'react';
import { mount, shallow } from 'enzyme';
import keycode from 'keycode';
import cases from 'jest-in-case';

import TreeViewItem, { getItemIcon } from './TreeViewItem.component';
import TreeView from '../TreeView.component';

const item = {
	id: 1,
	name: 'grandpa',
	actions: [
		{
			action: () => 'itemRemoveCallback',
			icon: 'talend-trash',
			label: 'remove element',
		},
	],
	children: [
		{
			id: 11,
			name: 'mami',
			toggled: true,
			children: [{ id: 111, name: 'me' }, { id: 112, name: 'bro' }],
		},
		{
			id: 12,
			name: 'aunt',
			toggled: false,
			children: [{ id: 121, name: 'cousin' }],
		},
	],
	toggled: true,
	counter: 101,
	showCounter: true,
};
const itemWithIcon = {
	name: 'grandpa',
	icon: {
		name: 'src-http://static.pokemonpets.com/images/monsters-images-300-300/106-Hitmonlee.png',
		title: 'A pokemon',
	},
	actions: [
		{
			action: () => 'itemRemoveCallback',
			icon: 'talend-trash',
			label: 'remove element',
		},
	],
	toggled: true,
	counter: 101,
	showCounter: true,
};
const items = [item, { id: 2, name: 'grandma' }, { id: 3, name: 'granduncle' }, itemWithIcon];

const defaultProps = {
	id: 'my-treeview',
	index: 1,
	item,
	itemSiblings: items,
	level: 2,
	onSelect: jest.fn(),
	onToggle: jest.fn(),
	onToggleAllSiblings: jest.fn(),
};
const propsWithIcons = {
	id: 'my-treeview',
	index: 1,
	item: itemWithIcon,
	itemSiblings: items,
	level: 2,
	onSelect: jest.fn(),
	onToggle: jest.fn(),
	onToggleAllSiblings: jest.fn(),
};

describe('getItemIcon', () => {
	it('should return given icon when toggled', () => {
		expect(getItemIcon('my', true)).toBe('my');
	});

	it('should return closed version of the given icon when not toggled', () => {
		expect(getItemIcon('my', false)).toBe('my-closed');
	});

	it('should return talend-folder icon when toggled without icon name', () => {
		expect(getItemIcon(undefined, true)).toBe('talend-folder');
	});

	it('should return talend-folder-closed icon when not toggled and without icon name', () => {
		expect(getItemIcon(undefined, false)).toBe('talend-folder-closed');
	});
});

describe('TreeView', () => {
	it('should render', () => {
		// when
		const wrapper = shallow(<TreeViewItem {...defaultProps} />);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should render items with icon', () => {
		// when
		const wrapper = shallow(<TreeViewItem {...propsWithIcons} />);

		// then
		expect(wrapper.getElement()).toMatchSnapshot();
	});

	it('should toggle item on toggle button click', () => {
		// given
		const props = {
			...defaultProps,
			onToggle: jest.fn(),
		};
		const wrapper = shallow(<TreeViewItem {...props} />);
		const event = { stopPropagation: jest.fn() };

		// when
		wrapper.find(`#${props.id}-toggle`).simulate('click', event);

		// then
		expect(event.stopPropagation).toBeCalled();
		expect(props.onToggle).toBeCalledWith(props.item);
	});

	it('should select item on click', () => {
		// given
		const props = {
			...defaultProps,
			onSelect: jest.fn(),
		};
		const wrapper = shallow(<TreeViewItem {...props} />);
		const event = { stopPropagation: jest.fn() };

		// when
		wrapper.simulate('click', event);

		// then
		expect(event.stopPropagation).toBeCalled();
		expect(props.onSelect).toBeCalledWith(props.item);
	});

	describe('gesture', () => {
		it('should select item on enter keydown', () => {
			// given
			const props = {
				...defaultProps,
				onSelect: jest.fn(),
			};
			const wrapper = shallow(<TreeViewItem {...props} />);
			const event = { keyCode: keycode.codes.enter, stopPropagation: jest.fn() };
			expect(props.onSelect).not.toBeCalled();

			// when
			wrapper.simulate('keydown', event);

			// then
			expect(event.stopPropagation).toBeCalled();
			expect(props.onSelect).toBeCalledWith(props.item);
		});

		it('should select item on space keydown', () => {
			// given
			const props = {
				...defaultProps,
				onSelect: jest.fn(),
			};
			const wrapper = shallow(<TreeViewItem {...props} />);
			const event = { keyCode: keycode.codes.space, stopPropagation: jest.fn() };
			expect(props.onSelect).not.toBeCalled();

			// when
			wrapper.simulate('keydown', event);

			// then
			expect(event.stopPropagation).toBeCalled();
			expect(props.onSelect).toBeCalledWith(props.item);
		});

		it('should toggle opened item on left keydown', () => {
			// given
			const props = {
				...defaultProps,
				onToggle: jest.fn(),
			};
			const wrapper = shallow(<TreeViewItem {...props} />);
			const event = { keyCode: keycode.codes.left, stopPropagation: jest.fn() };
			expect(props.onToggle).not.toBeCalled();

			// when
			wrapper.simulate('keydown', event);

			// then
			expect(event.stopPropagation).toBeCalled();
			expect(props.onToggle).toBeCalledWith(props.item);
		});

		it('should toggle closed item on right keydown', () => {
			// given
			const props = {
				...defaultProps,
				item: { ...defaultProps.item, toggled: false },
				onToggle: jest.fn(),
			};
			const wrapper = shallow(<TreeViewItem {...props} />);
			const event = { keyCode: keycode.codes.right, stopPropagation: jest.fn() };
			expect(props.onToggle).not.toBeCalled();

			// when
			wrapper.simulate('keydown', event);

			// then
			expect(event.stopPropagation).toBeCalled();
			expect(props.onToggle).toBeCalledWith(props.item);
		});

		it('should open all siblings on * keydown', () => {
			// given
			const props = {
				...defaultProps,
				onToggleAllSiblings: jest.fn(),
			};
			const wrapper = shallow(<TreeViewItem {...props} />);
			const event = { keyCode: keycode.codes['numpad *'], stopPropagation: jest.fn() };
			expect(props.onToggleAllSiblings).not.toBeCalled();

			// when
			wrapper.simulate('keydown', event);

			// then
			expect(event.stopPropagation).toBeCalled();
			expect(props.onToggleAllSiblings).toBeCalledWith(items);
		});

		function getSelector({ level, posinset }) {
			return `li[aria-level=${level}][aria-posinset=${posinset}]`;
		}

		function testFocus({ elementPosition, expectedActivePosition, keyCode }) {
			// given
			const wrapper = mount(
				<TreeView
					id="my-treeview"
					onToggle={jest.fn()}
					onToggleAllSiblings={jest.fn()}
					onSelect={jest.fn()}
					structure={items}
				/>,
			);
			const event = { keyCode };
			const element = wrapper.find(getSelector(elementPosition));
			const expectedActiveElementId = wrapper.find(getSelector(expectedActivePosition)).prop('id');

			// when
			element.simulate('keydown', event);

			// then
			expect(document.activeElement.getAttribute('id')).toBe(expectedActiveElementId);
		}

		/**
		 * The current data gives this tree. Legend : (level, poinset)
		 * > grandpa (1, 1)
		 *      > mami (2, 1)
		 *           me (3, 1)
		 *           bro (3, 2)
		 *      > aunt (2, 2)
		 *   grandma (1, 2)
		 *   ganduncle (1, 3)
		 * > withIcon (1, 4)
		 */
		cases('focus', testFocus, [
			{
				name: 'should focus its parent on left keydown',
				elementPosition: { level: 3, posinset: 1 },
				expectedActivePosition: { level: 2, posinset: 1 },
				keyCode: keycode.codes.left,
			},
			{
				name: "should focus opened item's first child on right keydown",
				elementPosition: { level: 1, posinset: 1 },
				expectedActivePosition: { level: 2, posinset: 1 },
				keyCode: keycode.codes.right,
			},
			{
				name: 'should focus next item on down keydown',
				elementPosition: { level: 2, posinset: 2 },
				expectedActivePosition: { level: 1, posinset: 2 },
				keyCode: keycode.codes.down,
			},
			{
				name: 'should focus next item on down keydown',
				elementPosition: { level: 2, posinset: 2 },
				expectedActivePosition: { level: 1, posinset: 2 },
				keyCode: keycode.codes.down,
			},
			{
				name: 'should focus previous item on up keydown',
				elementPosition: { level: 2, posinset: 2 },
				expectedActivePosition: { level: 3, posinset: 2 },
				keyCode: keycode.codes.up,
			},
			{
				name: 'should focus first item on home keydown',
				elementPosition: { level: 2, posinset: 2 },
				expectedActivePosition: { level: 1, posinset: 1 },
				keyCode: keycode.codes.home,
			},
			{
				name: 'should focus last item on end keydown',
				elementPosition: { level: 2, posinset: 2 },
				expectedActivePosition: { level: 1, posinset: 4 },
				keyCode: keycode.codes.end,
			},
		]);
	});
});
