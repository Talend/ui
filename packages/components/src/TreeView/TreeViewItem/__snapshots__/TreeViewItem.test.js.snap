// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeView item should render 1`] = `
<li
  aria-expanded={false}
  aria-level={2}
  aria-posinset={1}
  aria-selected={false}
  aria-setsize={4}
  className="tc-treeview-item-li theme-tc-treeview-li"
  id="my-treeview"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="treeitem"
  tabIndex={-1}
>
  <div
    className="tc-treeview-item theme-tc-treeview-item"
    style={
      Object {
        "paddingLeft": "62px",
      }
    }
  >
    <Action
      aria-hidden={true}
      className="theme-tc-treeview-toggle"
      icon="talend-caret-down"
      iconTransform="rotate-270"
      id="my-treeview-toggle"
      label=""
      link={true}
      onClick={[Function]}
      tabIndex="-1"
    />
    <TreeViewIcon
      isOpened={false}
    />
    <span
      className="tc-treeview-item-name theme-tc-treeview-item-name"
    >
      grandpa
    </span>
    <div
      className="theme-tc-treeview-item-ctrl"
    >
      <Translate(Badge)
        label="101"
      />
      <Action
        hideLabel={true}
        icon="talend-trash"
        id="my-treeview-talend-trash"
        label="remove element"
        link={true}
        onClick={[Function]}
        tooltipPlacement="right"
      />
    </div>
  </div>
</li>
`;

exports[`TreeView item should render items with icon 1`] = `
<li
  aria-level={2}
  aria-posinset={1}
  aria-selected={false}
  aria-setsize={4}
  className="tc-treeview-item-li theme-tc-treeview-li"
  id="my-treeview"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="treeitem"
  tabIndex={-1}
>
  <div
    className="tc-treeview-item theme-tc-treeview-item"
    style={
      Object {
        "paddingLeft": "62px",
      }
    }
  >
    <TreeViewIcon
      icon={
        Object {
          "name": "src-http://static.pokemonpets.com/images/monsters-images-300-300/106-Hitmonlee.png",
          "title": "A pokemon",
        }
      }
      isOpened={false}
    />
    <span
      className="tc-treeview-item-name theme-tc-treeview-item-name"
    >
      grandpa
    </span>
    <div
      className="theme-tc-treeview-item-ctrl"
    >
      <Translate(Badge)
        label="101"
      />
      <Action
        hideLabel={true}
        icon="talend-trash"
        id="my-treeview-talend-trash"
        label="remove element"
        link={true}
        onClick={[Function]}
        tooltipPlacement="right"
      />
    </div>
  </div>
</li>
`;
